### ISSUE TITLE
Performance issues - how to scale ?

### ISSUE THREAD

**Issue by psadovsky** [openfga/openfga#445]
I would like to describe our setup of OpenFGA with PostgreSQL. We deployed OpenFGA into Azure AKS, the PostgreSQL is separated service managed by Azure, so we did not deploy it in the same pod with OpenFGA. I guess there is some overhead regarding network communication. The following picture describing our setup:

![image](https://user-images.githubusercontent.com/72700106/210647754-eeb9784a-495f-4bb1-92e9-5ce305739cb5.png)

Our frontend needs to authorize multiple relations at one request. For instance the biggest request contains 16 relations to authorize. Frontend calls the permission service and I have to create 16 parallel requests for the best performance option. Unfortunately, it took 5-6 seconds, I cannot imagine how OpenFGA can handle thousands or hundred of thousands requests parallely. We have already started to cache checking relations, but the performance of OpenFGA is weak in general.

The question is what we do wrong, if anything ? or the problem is in maturity of the project ?

**Comment by aaguiarz** [comment]
Hi @psadovsky,

Could you please provide us additional context? 

- What version of OpenFGA are you using?
- Are you calling Check() or ListObjects()?
- How many tuples do you have?
- Are you deploying the permission service, OpenFGA and PostgreSQL in the same Azure region? In the same VPC?
- Can you share the code you are using to make the 16 parallel requests? Are you sure the requests are not being serialized? 
- Did you run the same test in a different environment (e.g. your local machine)? If so, what results did you get?
- Did you bootstrap the PostgresSQL store with the `openfga migrate` utility, and if not, have you indexed the tables appropriately?

Thanks a lot

**Comment by psadovsky** [comment]
Hello @aaguiarz,

Thank you for your prompt response. Step by step:

- v0.3.1
- check()
- 404
- Yes
- It will be at the bottom of this list and YES, it is serialized
- No, we have not already run those tests in a different environment, but that is the plan.
- Yes, we used "openfga migrate", tables are properly indexed.
- from the backend

```java
    @Override
    public Flux<RelationCheckResponseDTO> checkRelations(String storeId, List<RelationCheckRequestDTO> checkRequests) {

        return Flux.fromIterable(checkRequests)
                .flatMapSequential((checkRequest) -> getCachedOrSendRequest(storeId, checkRequest));
    }

    private Mono<RelationCheckResponseDTO> getCachedOrSendRequest(String storeId, RelationCheckRequestDTO relationCheckRequestDTO) {

        Cache relationsCache = cacheManager.getCache("relations");
        Optional<RelationCheckResponseDTO> resultOptional;
        if (relationsCache != null) {
            TupleKeyDTO tupleKeyDTO = relationCheckRequestDTO.getTupleKey();
            String cacheKey = OpenFgaUtil.buildCacheKey(tupleKeyDTO.getUser(), tupleKeyDTO.getRelation(), tupleKeyDTO.getObject());
            resultOptional = Optional.ofNullable(relationsCache.get(cacheKey, RelationCheckResponseDTO.class));

        return resultOptional
                .map(Mono::just)
                .orElseGet(() -> checkRelationRequest(storeId, relationCheckRequestDTO)
                        .doOnNext((response) -> relationsCache.put(cacheKey, response)));
        }
        return checkRelationRequest(storeId, relationCheckRequestDTO);
    }

    private Mono<RelationCheckResponseDTO> checkRelationRequest(String storeId, RelationCheckRequestDTO relationCheckRequestDTO) {

        if (StringUtils.hasText(storeId) && relationCheckRequestDTO != null) {
            return webClient.post()
                    .uri(uriBuilder -> uriBuilder
                            .path(ApiPaths.STORES_PATH)
                            .path(ApiPaths.STORES_ID_PATH_VAR)
                            .path(ApiPaths.CHECK_PATH)
                            .build(storeId))
                    .body(Mono.just(relationCheckRequestDTO), RelationCheckRequestDTO.class)
                    .retrieve()
                    .bodyToMono(RelationCheckResponseDTO.class)
                    .onErrorResume(WebClientResponseException.class, e -> {
                        if (HttpStatus.BAD_REQUEST.equals(e.getStatusCode()) &&
                                OpenFgaErrorCode.VALIDATION_ERROR.equals(e.getStatusText())) {
                            return Mono.just(RelationCheckResponseDTO.builder().allowed(Boolean.FALSE).build());
                        }
                        return Mono.error(e);
                    });
        }
        throw new BadRequestApiException(CommonErrorCode.BUSINESS_VALIDATION_FAILED,
                "Relation cannot be checked without store id and relation tuple.");
    }
```

**Comment by aaguiarz** [comment]
Hi @psadovsky 

If the requests are not being sent in parallel, then 6s/16req = 375ms/req. That's quite slow. If indexes are there, then the next suspect would be latency between different services. 

Another option would be that the model is really complex with a lot of nested levels/recursion. I don't think that would cause this, but it might be worth checking. If you can share your model that would be awesome. If you prefer to send it to me directly, please email it to andres.aguiar at okta.

**Comment by aaguiarz** [comment]
Hi @psadovsky 

If you can also provide us the tuples + the check() you run, we can test it on our end.

Thanks

**Comment by psadovsky** [comment]
> 

@aaguiarz I send you an email.

**Comment by psadovsky** [comment]
> Hi @psadovsky
> 
> If you can also provide us the tuples + the check() you run, we can test it on our end.
> 
> Thanks

Can I give you SQL scripts? Does" the check() you run" mean giving you the client code?

**Comment by psadovsky** [comment]
We set the value of "--datastore-max-idle-conns 30" and it helped to cut the time by half. However, it still does not work how we would expect. To be honest, without our own caching, it would not be useful.

I tried to do a simple performance test on my laptop (it is a weak machine, but the tested load is not so heavy):

![image](https://user-images.githubusercontent.com/72700106/212563447-5d14efde-94e2-4ad1-b33c-e62b76b89c0d.png)

Note: I switched off custom caching.

20 threads/users/requests calling our permission service which transforms the individual request into OpenFGA "check" request, which means generating 16 parallel OpenFGA check requests for one JMeter request. Together it gives 16 x 20 = 320 requests. One JMeter request took approximately 2 seconds. Except that I set --datastore-max-idle-conns  100.

**Comment by craigpastro** [comment]
👋 We have now released v0.3.3, which we believe will help this issue. (See also #380.) Please try it out and let us know how it goes.

**Comment by psadovsky** [comment]
@craigpastro I have good news, it looks like it helped a lot 
![image](https://user-images.githubusercontent.com/72700106/216011167-adbe811a-e6a4-4e14-afef-29c0b8b0a56e.png)

We went down from approx. 2 seconds to 250ms. Great job 🥳

**Comment by jon-whit** [comment]
@psadovsky can we close this at this point then? If anything else comes up you can submit another issue 👍

**Comment by psadovsky** [comment]
Yes, we can. Thanks 👍
